### ISSUE TITLE
Single line of output to Find-result window for lines with 1+ hits

### ISSUE THREAD

**Issue by DanielTheGreat** [notepad-plus-plus/notepad-plus-plus#1499]
I don't use it often, but I like Notepad++ for the occasional coding (hobby) project!  Just one little thing slightly annoys me, so I wondered whether others would agree to a small change to the way the  'Find All in...' feature operates...

I sometimes use 'Find All in All Opened Documents' (or the other one) to find all occurrences of a particular string that I want to check and/or edit.  But if I search for a string (perhaps a variable or register name) that occurs twice (or more) in one line, the search results pane shows that line twice (or more).  This simply clutters the search results, and seems to serve no useful purpose.

Here's an example:
# define    vca25    vca2_addr.bit.b5    // Voltage detection 0 enable bit

If I'm searching for 'vca2' to find everywhere the register is accessed, I don't need to see that line appear twice in the results;  I just want to know every file and every line where that register appears.

Would it be practical (and acceptable to most users) to change the 'Find All' code such that when a string is identified in a particular file/line, **that** line is then skipped from further searching so it appears only once in the results pane?

I realise this will complicate the code a little, since I presume there is much common code between 'Find All in...' and 'Replace All in...' functions, whereas for 'Replace All in...' you **do** want every instance found, not just the first per line!

Thanks, Daniel
(Gerroa, Australia)

**Comment by MetaChuh** [comment]
please update your notepad++ to the latest version [<kbd>Notepad++ 7.6.6</kbd>](https://notepad-plus-plus.org/download/v7.6.6.html) and migrate or reinstall your plugins if you update from version 7.5.9 or below.
if your issue still persists, and it is still of relevance to you, please feel free to file a new issue as easy to understand and reproducible as possible.

thank you for your support and best regards.

-----
```
general notes:

please close your issues as soon as you are satisfied, to help
keeping the notepad++ issue tracker clean.

when creating a new issue, please provide all information as shown at the issue template.
your notepad++ debug info is part of the minimum requirements.
(? > Debug Info... > Copy debug info into clipboard).

please also conduct a search at the issue tracker beforehand, to check if a similar issue
already exists.

for general questions, or if it is not sure, whether your issue is directly related to
the notepad++ source code, please visit us at the notepad++ community forum, and
search if related topics exists. you are welcome to post either at similar
topics, or to create a new topic at Help Wanted or General Discussion.
(no extra account is needed, just use your github account to sign in)
```
[<p align="center"><kbd> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; Click here to visit the Notepad++ Community Forum &emsp; &emsp; &emsp; &emsp; &emsp; &emsp;&emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; </kbd></p>](https://notepad-plus-plus.org/community/)

**Comment by sasumner** [comment]
So this issue will still exist in newer Notepad++.  However, why not conduct your search in Regular Expression search mode like this:

    (?-s)vca2.*

or like this with the `. matches newline` box unticked:

    vca2.*

**Comment by DanielTheGreat** [comment]
I'm not yet conversant with Regular Expressions; indeed just earlier today I printed out many pages from a 'manual', to start learning how to use them.  My RE familiarity notwithstanding, it's obviously more work (mental and physical!)  to do a RE search than just highlighting some text and searching for it.

I still don't understand what advantage is gained by showing the same line twice just because it contains two instances of the search string.  Why not just highlight the two instances and show the line once?

**Comment by sasumner** [comment]
>I still don't understand what advantage is gained by showing the same line twice just because it contains two instances of the search string. Why not just highlight the two instances and show the line once?

I'm sorry you don't see the utility.  For some of us, the way it works is the desirable way (and in fact in their opinions is the ONLY way. :) )

So the regex solution provides a compromise.  It is up to you to decide if you can bear using it or not. :)

**Comment by DanielTheGreat** [comment]
I understand the power of regular expression, which is why I decided earlier to try to learn them.  And thanks for your specific example, which works.  But it's a big task for me to learn, and will take me time.

I can't force anybody to make changes, and I do greatly appreciate the work that has gone into making this powerful program, and providing it free of charge.  But just to satisfy my curiosity and not leave me perplexed, would you please describe a scenario where having a line appear twice in a search result because it has two instances of the string would bring an advantage over showing that line just once?

**Comment by sasumner** [comment]
@DanielTheGreat It's really just a philosophy.  If you haven't understood it already you probably won't (but that is in no way a put-down).  Everyone tends to use tools differently.  I certainly understand your philosophy, that's why I long ago put the little regex trick (which really doesn't require a big "learning regex" effort, just a quick add `.*` to the end of a search and tick regex search mode) in my "toolbag" for when I want that functionality (rare).

But I'll give you one:  So one of the features of the Find result panel is that double-clicking an entry (i.e. a line) there (keyboard way exists as well) will jump your caret to the match.  If there are multiple matches per line, how would this work?  Sure you could double-click, look at the match, and then press Find Next to see if there are additional matches per line, but this would be laborious if in most cases there are not an additional match(es). Plus it is something you have to remember to check.  For me, I'd get mentally involved a thinking about the match it found (the first one) and sometimes forget to consider the rest of the line's possible matches.

**Comment by DanielTheGreat** [comment]
I can sort of see your point in that example, but surely the same result could be achieved by highlighting each instance of a hit in the one line with the same 'jump-to' functionality?  Furthermore, when you jump to the first instance of a hit in a line, all the others are highlighted anyway (see screen-shot), so you know how many are in the line that needs 'fixing'.
(Not sure if the screen-shot has been attached/included – the editing screen shows no confirmation after dragging-dropping it.)

**Comment by DanielTheGreat** [comment]
Well, no sign of my .png screen-shot, so I don't know how to upload it. :-(

**Comment by sasumner** [comment]
>with the same 'jump-to' functionality

Not sure I'm following the meaning here...

> all the others are highlighted anyway

Are you sure you've considered the case of really LONG lines? Where not all matches per line could be visible in the window width...

>no sign of my .png screen-shot, so I don't know how to upload it

Copy and then paste here works (or should)

**Comment by DanielTheGreat** [comment]
"with the same 'jump-to' functionality" -- you have ONE lines with EACH of the multiple search strings highlighted in it.  Double-clicking EACH highlighted string (in turn) would jump to that part of the line, as it currently does when double-clicking a highlighted search result string.

"Are you sure you've considered the case of really LONG lines?" -- I would consider that poor programming style! :-)  I always break unavoidably long lines so that each part is visible in the pane width I commonly use.  I believe having to scroll horizontally to read fragments of a line lowers comprehension.  (Just my untrained opinion! :-)

Still can't paste screen-shots into my posts.  Maybe the host software just doesn't like Firefox 56?  (I refuse to use the crippled 'Quantum' versions.)

**Comment by DanielTheGreat** [comment]
PS: I'm happy to close this thread any time you've had enough backchat!

**Comment by sasumner** [comment]
>I would consider that poor programming style

Who's talking about programming?  Notepad++ is a text editor.  All types of text (e.g. in this case, text files with really long lines) are valid scenarios.  I certainly program as well, but a lot of my text manipulation needs are in files with long lines, where multiple matches would not be human-visible simultaneously (horizontally).

>I'm happy to close this thread any time you've had enough backchat!

Nono...it is often that extended issue discussion brings out unthought of aspects.  Someone that is considering implementing a feature then gets a better insight into how people use the product and the proposed feature.

**Comment by DanielTheGreat** [comment]
Sorry, my blinkered vision!  I consider Notepad++ as a "programmer's text editor", but of course you're right, and can be used in non-programming scenarios too.

Mind you, as a former newsletter editor, I would also consider a plain text file with past-the-margin lines to be 'poor style' too.  (Though I'm sure you'll come up with an instance where it is justified! :-)  But if I encountered one, I would simply enable line-wrap, which is usually the default for plain (non-programming) text editors.

**Comment by sasumner** [comment]
Sometimes it is not about "style", it is about machine-created text needing to be searched/modified to be sent to some other machine process.  It really isn't meant to be human-nice (e.g. wrapped so it is "pretty").  Think about DNA sequencing (although that isn't my application).  In fact, the columns that my data is restricted to is important, so a wrapped view would really hinder rather than assist.

So I'm not trying to refute any point you come up with.  Just trying to point out that there are all kinds of text processing needs, not all of them involving "appearance / style".  And since a simple regex can aid and provide the 2 different functionalities, I say "why not take advantage of it?".  Especially since N++ developers have long been aware of this "issue" (I have seen it requested a few times) and have chosen to leave it alone, well...  

Cheers, good discussions...but I think we've tapped out.  :)

**Comment by DanielTheGreat** [comment]
Yes, thanks for broadening my view of what people use text editors for.  Thanks too for the RegEx tips – I'm starting on the RegEx manual today!

**Comment by sasumner** [comment]
This feature-request is REJECTED by the author of Notepad++ ( https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9088#issuecomment-720016310 )
It will not be reconsidered.
Please do not open any new issues on this topic as they will be immediately closed.
The suggestion is to use this workaround:
* Set your _Search mode_ to _Regular expression_.
* Add the following to the end of your _Find what_ expression:  `(?-s).*`

Example: To show only the first hit on a line for `foo`, make your _Find what_ expression `foo(?-s).*`

**Comment by sasumner** [comment]
One more person expressing a wish for this feature:  https://community.notepad-plus-plus.org/topic/20384/duplicate-same-lines-with-regex-search

Same posting has a nice demo (with a picture) of my proposed workaround. :-)

**Comment by sasumner** [comment]
One more person expressing a wish for this feature:
![image](https://user-images.githubusercontent.com/30118311/106751327-06c78a80-65f7-11eb-8415-41895af1d004.png)

**Comment by sasumner** [comment]
Another:  https://community.notepad-plus-plus.org/topic/21193/stop-searching-a-line-after-one-match
